### PR TITLE
Jupyter-notebook for the generators module

### PR DESCRIPTION
**Pull request by cndolo**
This PR adds a new jupyter-notebook for the `generators` module. I made minor changes to the documentation too.

### CODE REVIEW

File: notebooks/Generators.ipynb [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Generate sequences\n",
+    "lfr.generatePowerlawDegreeSequence(20, 50, -2);\n",
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "lfr.generatePowerlawDegreeSequence(20, 50, -2)\n",
```

File: notebooks/Generators.ipynb [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+    "lfr.generatePowerlawCommunitySizeSequence(10, 50, -1);\n",
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "lfr.generatePowerlawCommunitySizeSequence(10, 50, -1)\n",
```

File: notebooks/Generators.ipynb [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Verify\n",
+    "print(lfrG.numberOfNodes())"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "print(lfrG.numberOfNodes(), lfrG.numberOfEdges())"
```

File: notebooks/Generators.ipynb [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+    "LFR benchmark is an algorithm that generates benchmark networks. The node degrees are distributed according to a power law with different exponents. \n",
+    "\n",
+    "The [LFR(n)](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=lfr#networkit.generators.LFRGenerator) constructor only expects the number of nodes the generated graph should have. However, before generating the graph one needs to set a [degree sequence](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=degreese#networkit.generators.LFRGenerator.setDegreeSequence), [community size sequence](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=communitysi#networkit.generators.LFRGenerator.setCommunitySizeSequence) and [ mu](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=setm#networkit.generators.LFRGenerator.setMu) or generate the sequences using the provided [generate-methods](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=generatepower#networkit.generators.LFRGenerator.generatePowerlawCommunitySizeSequence). \n",
+    "\n",
+    "For this example, we shall generate the sequences."
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "For this example, we generate the sequences a power-law degree sequence with average degree 20, maximum degree 50, and exponent of the node degree distribution -2. We also generate a power-law community size sequence  with minimum size 10, maximum size 50, and exponent of the community size distribution -2. Finally, we set the mixing parameter `mu` to 0.5."
```

File: notebooks/Generators.ipynb [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Initialize algorithm\n",
+    "crg = nk.generators.ClusteredRandomGraphGenerator(100,10,0.5,0.01)"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "crg = nk.generators.ClusteredRandomGraphGenerator(100, 10, 0.5, 0.01)"
```

File: notebooks/Generators.ipynb [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The clustered random graph generates a clustered random graph. The number of nodes and the number of edges are adjustable as well as the probabilities for intra-cluster and inter-cluster edges.\n",
+    "\n",
+    "The constructor [ClusteredRandomGraphGenerator(n, k, pin, pout)](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=clustered#networkit.generators.ClusteredRandomGraphGenerator) expects the number of nodes `n` and the number of clusters `k` followed by the intra-cluster edge probability and the inter-cluster edge probability as `pin` and `pout` respectively.\n",
+    "\n",
+    "A graph with 100 nodes grouped in 10 clusters can be generated as follows:"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "A graph with 100 nodes grouped in 10 clusters with intra-cluster edge probability 0.5 and inter-cluster edge probability 0.01 can be generated as follows:"
```

File: notebooks/Generators.ipynb [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Verify\n",
+    "print(crgG.numberOfNodes())"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "print(crgG.numberOfNodes(), crgG.numberOfEdges())"
```

File: notebooks/Generators.ipynb [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Verify\n",
+    "print(dmgG.numberOfNodes())"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "print(dmgG.numberOfNodes(), dmgG.numberOfEdges())"
```

File: notebooks/Generators.ipynb [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Generate degree sequence\n",
+    "degSeq = [4,3,2,1,1,1]"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "degSeq = [4, 3, 2, 1, 1, 1]"
```

File: notebooks/Generators.ipynb [hunk 10]
```diff
@@ -0,0 +1,580 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Given an arbitrary degree sequence, the Chung-Lu generative model will produce a random graph with the same expected degree sequence if possible.\n",
+    "\n",
+    "The constructor [ChungLuGenerator(degreeSequence)](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=chung#networkit.generators.ChungLuGenerator) expects a degree sequence as a parameter.\n",
+    "\n",
+    "In order to create a graph with 5 nodes, we first need to generate a degree sequence that will be passed to the constructor."
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "In order to create a graph with 5 nodes, we first need to generate a degree sequence that will be passed to the constructor. Note that the degree sequence is not required to be sorted."
```

File: notebooks/Generators.ipynb [hunk 11]
```diff
@@ -0,0 +1,580 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Verify\n",
+    "chungLuSeq = []\n",
+    "for u in range (clgG.upperNodeIdBound()):\n",
+    "    chungLuSeq.append(clgG.degree(u))\n",
+    "\n",
+    "print(clgG.numberOfNodes())"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "print(clgG.numberOfNodes(), clgG.numberOfEdges())"
```

You might also print out the actual degree sequence.

File: notebooks/Generators.ipynb [hunk 12]
```diff
@@ -0,0 +1,580 @@
...
+    "## Havel-Hakimi Generator "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Havel-Hakimi algorithm for generating a graph according to a given degree sequence $(d_1, d_2,...,d_n)$. The degree sequence must be non-increasing, i.e., $d_1$ must be the highest degree. \n",
+    "\n",
+    "The contructor [HavelHakimiGenerator(sequence, ignoreIfRealizable=True)](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=havel#networkit.generators.HavelHakimiGenerator) expects the degree sequence as a mandatory parameter. If `ignoreIfRealizable` is true, the graph is generated even if the degree sequence is not realizable. Some nodes may then get lower degrees than requested in the sequence. If If `ignoreIfRealizable` is false and the sequence is not realizable, an exception is thrown and the graph cannot be generated."
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "The contructor [HavelHakimiGenerator(sequence, ignoreIfRealizable=True)](https://networkit.github.io/dev-docs/python_api/generators.html?highlight=havel#networkit.generators.HavelHakimiGenerator) expects the degree sequence as a mandatory parameter. If `ignoreIfRealizable` is true, the graph is generated even if the degree sequence is not realizable. Some nodes may then get lower degrees than requested in the sequence. If `ignoreIfRealizable` is false and the sequence is not realizable, an exception is thrown and the graph cannot be generated."
```

File: notebooks/Generators.ipynb [hunk 13]
```diff
@@ -0,0 +1,580 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Verify\n",
+    "print(hhgG.numberOfNodes())\n",
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "print(hhgG.numberOfNodes(), hhgG.numberOfEdges())\n",
```

File: notebooks/Generators.ipynb [hunk 14]
```diff
@@ -0,0 +1,580 @@
...
+    "for u in range (hhgG.upperNodeIdBound()):\n",
+    "    print(\"Degree sequences are equal: \", (sequence[u] == hhgG.degree(u)))"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "    assert(sequence[u] == hhgG.degree(u))"
```

File: notebooks/Generators.ipynb [hunk 15]
```diff
@@ -0,0 +1,580 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Verify \n",
+    "print(mgG.numberOfNodes())\n",
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
    "print(mgG.numberOfNodes(), mgG.numberOfEdges())\n",
```

File: notebooks/Generators.ipynb [hunk 16]
```diff
@@ -0,0 +1,580 @@
...
+    "print(mgG.numberOfEdges())"
```
Comment (angriman) [inline notebooks/Generators.ipynb]:
```suggestion
```

**Comment by cndolo** [comment]
Thanks for the review @angriman . i have addressed the suggestions.

File: notebooks/Generators.ipynb (hunk 11 above)
Comment (angriman) [reply notebooks/Generators.ipynb]: This is not resolved yet.

**Comment by angriman** [comment]
Thank you, there are still some comments to be resolved, I let them open.

**Pull request merged**